crates/cargo-contract/src/cmd/extrinsics/balance.rs

```diff
@@ -0,0 +1,289 @@
+// Copyright 2018-2022 Parity Technologies (UK) Ltd.
+// This file is part of cargo-contract.
+//
+// cargo-contract is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// cargo-contract is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with cargo-contract.  If not, see <http://www.gnu.org/licenses/>.
+
+use std::{
+    fmt::Display,
+    str::FromStr,
+};
+
+use rust_decimal::{
+    self,
+    Decimal,
+};
+use rust_decimal_macros::dec;
+use serde_json::json;
+
+use super::{
+    Balance,
+    Client,
+};
+
+use anyhow::{
+    Context,
+    Ok,
+    Result,
+};
+
+/// Represents different formats of a balance
+#[derive(Debug, Clone)]
+pub enum BalanceVariant {
+    /// Default format: no symbol, no denomination
+    Default(Balance),
+    /// Denominated format: symbol and denomination are present
+    Denominated(String),
+}
+
+#[derive(Debug, Clone)]
+pub struct TokenMetadata {
+    /// Number of denomination used for denomination
+    pub denomination: u128,
+    /// Token symbol
+    pub symbol: String,
+}
+
+impl TokenMetadata {
+    /// Query [TokenMetadata] through the node's RPC
+    pub async fn query(client: &Client) -> Result<Self> {
+        let sys_props = client.rpc().system_properties().await?;
+
+        let default_decimals = json!(12);
+        let default_units = json!("UNIT");
+        let decimals = sys_props
+            .get("tokenDecimal")
+            .unwrap_or(&default_decimals)
+            .as_u64()
+            .context("error converting decimal to u64")?;
+        let symbol = sys_props
+            .get("tokenSymbol")
+            .unwrap_or(&default_units)
+            .as_str()
+            .context("error converting symbol to string")?;
+        let denomination: u128 = format!("1{}", "0".repeat(decimals as usize)).parse()?;
+        Ok(Self {
+            denomination,
+            symbol: symbol.to_string(),
+        })
+    }
+}
+
+impl BalanceVariant {
+    /// Converts BalanceVariant into Balance.
+    ///
+    /// Throws Error if [BalanceVariant::Denominated(String)] is in an incorrect format.
+    pub fn denominate_balance(&self, token_metadata: &TokenMetadata) -> Result<Balance> {
+        match self {
+            BalanceVariant::Default(balance) => Ok(*balance),
+            BalanceVariant::Denominated(input) => {
+                if let Some(balance_str) =
+                    input.strip_suffix(&format!("k{}", token_metadata.symbol))
```

Comment:
Too much repetition in this method, making the calculations unclear. We should separate parsing from the calculations by doing it up front in a `FromStr` impl. You could flatten it out to something like:

```
#[derive(Debug, Clone)]
pub struct Balance {
    value: u128,
    unit: Option<UnitPrefix>,
    symbol: Option<String>,
}

#[derive(Debug, Clone)]
pub enum UnitPrefix {
    Giga,
    Mega,
    Kilo,
    One,
    Milli,
    Micro,
    Nano,
}
```

Reply:
I agree that we can structure denominated balance. However, I think there should be two struct `Balance` and `DenominatedBalance` because they are of different types: `DenominatedBalance` is represented as decimal, it has a a point like `45.234234MDOT` whereas `Balance` is of `u128` type